File: server/services/leaderboard.js
/**
 * Leaderboard Service
 * ====================
 * Sharded leaderboard engine using Redis sorted sets.
 *
 * Architecture for 1M+ players:
 * - Scores distributed across N shards (consistent hashing by playerId)
 * - Shard merging for top-K queries
 * - Hot cache for frequently accessed pages
 * - Background refresh for stale shards
 *
 * Redis key structure:
 *   stem:lb:{tenantId}:{gameId}:shard:{n}  → ZSET (score → playerId)
 *   stem:lb:{tenantId}:global:shard:{n}    → ZSET (totalScore → playerId)
 *   stem:lb_result:{...}                   → Cached JSON result
 */

const cache = require('./cache');
const db = require('../models/db');
const config = require('../config');

const SHARD_COUNT = config.leaderboard.shardCount;

/**
 * Update a player's score in the sharded leaderboard.
 */
async function updateScore(tenantId, gameId, playerId, score) {
    const shard = _shardFor(playerId);

    // Per-game leaderboard
    const gameKey = `lb:${tenantId}:${gameId}:shard:${shard}`;
    const existing = await cache.zscore(gameKey, playerId);
    if (!existing || score > parseFloat(existing)) {
        await cache.zadd(gameKey, score, playerId);
    }
    await cache.expire(gameKey, 7200); // 2 hour TTL

    // Invalidate cached results for this game
    await _invalidateResultCache(tenantId, gameId);
}

/**
 * Update a player's global total score.
 */
async function updateGlobalScore(tenantId, playerId, totalScore) {
    const shard = _shardFor(playerId);
    const key = `lb:${tenantId}:global:shard:${shard}`;
    await cache.zadd(key, totalScore, playerId);
    await cache.expire(key, 7200);
}

/**
 * Get top-K entries by merging all shards.
 * This is O(K * SHARD_COUNT) which is efficient for reasonable K values.
 */
async function getTopK(tenantId, gameId, k = 50) {
    const candidates = [];

    // Gather top-K from each shard
    for (let s = 0; s < SHARD_COUNT; s++) {
        const key = `lb:${tenantId}:${gameId}:shard:${s}`;
        const entries = await cache.zrevrange(key, 0, k - 1, true);
        if (entries) {
            for (const entry of entries) {
                candidates.push({
                    playerId: entry.value,
                    score: entry.score
                });
            }
        }
    }

    // Sort merged candidates and take top K
    candidates.sort((a, b) => b.score - a.score);
    return candidates.slice(0, k);
}

/**
 * Get a player's approximate rank by counting higher scores across shards.
 */
async function getApproxRank(tenantId, gameId, playerId) {
    const shard = _shardFor(playerId);
    const key = `lb:${tenantId}:${gameId}:shard:${shard}`;

    const score = await cache.zscore(key, playerId);
    if (score === null) return null;

    // Count players with higher scores across all shards
    let higherCount = 0;
    for (let s = 0; s < SHARD_COUNT; s++) {
        const shardKey = `lb:${tenantId}:${gameId}:shard:${s}`;
        // Count entries with score > player's score
        const count = await cache.zcard(shardKey);
        if (count > 0) {
            // Approximate: assume uniform distribution within shard
            const topEntries = await cache.zrevrange(shardKey, 0, 0, true);
            if (topEntries && topEntries.length > 0 && topEntries[0].score > score) {
                // Binary estimate of how many are above
                higherCount += Math.round(count * 0.5); // rough approx
            }
        }
    }

    return higherCount + 1;
}

/**
 * Rebuild leaderboard cache from database.
 * Run periodically or on cache miss.
 */
async function rebuildFromDB(tenantId, gameId) {
    const result = await db.query(`
        SELECT player_id, high_score
        FROM game_progress
        WHERE tenant_id = $1 AND game_id = $2 AND high_score > 0
        ORDER BY high_score DESC
        LIMIT 10000
    `, [tenantId, gameId]);

    for (const row of result.rows) {
        const shard = _shardFor(row.player_id);
        const key = `lb:${tenantId}:${gameId}:shard:${shard}`;
        await cache.zadd(key, parseInt(row.high_score, 10), row.player_id);
        await cache.expire(key, 7200);
    }

    console.log(`Leaderboard rebuilt: ${tenantId}/${gameId} (${result.rows.length} entries)`);
}

// =========================================
// Internal
// =========================================

function _shardFor(playerId) {
    let hash = 0;
    for (let i = 0; i < playerId.length; i++) {
        hash = ((hash << 5) - hash) + playerId.charCodeAt(i);
        hash |= 0;
    }
    return Math.abs(hash) % SHARD_COUNT;
}

async function _invalidateResultCache(tenantId, gameId) {
    // Invalidate common cached result keys
    const periods = ['all', 'daily', 'weekly', 'monthly'];
    for (const period of periods) {
        await cache.del(`lb_result:${tenantId}:${gameId}:${period}:0:50`);
    }
}

module.exports = {
    updateScore,
    updateGlobalScore,
    getTopK,
    getApproxRank,
    rebuildFromDB
};

File: server-rs/src/services/leaderboard.rs
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

use crate::cache::Cache;
use crate::error::AppResult;

fn shard_index(player_id: &str, shard_count: u32) -> u32 {
    let mut hasher = DefaultHasher::new();
    player_id.hash(&mut hasher);
    let hash = hasher.finish();
    (hash % shard_count as u64) as u32
}

fn lb_key(tenant_id: &str, game_id: &str, shard: u32) -> String {
    format!("lb:{}:{}:shard:{}", tenant_id, game_id, shard)
}

fn global_key(tenant_id: &str, shard: u32) -> String {
    format!("lb:{}:global:shard:{}", tenant_id, shard)
}

pub async fn update_score(
    cache: &Cache,
    tenant_id: &str,
    game_id: &str,
    player_id: &str,
    score: f64,
    shard_count: u32,
) {
    let shard = shard_index(player_id, shard_count);
    let key = lb_key(tenant_id, game_id, shard);
    cache.zadd(&key, player_id, score).await;
    cache.expire(&key, 7200).await;
}

pub async fn update_global_score(
    cache: &Cache,
    tenant_id: &str,
    player_id: &str,
    total_score: f64,
    shard_count: u32,
) {
    let shard = shard_index(player_id, shard_count);
    let key = global_key(tenant_id, shard);
    cache.zadd(&key, player_id, total_score).await;
    cache.expire(&key, 7200).await;
}

pub async fn get_top_k(
    cache: &Cache,
    tenant_id: &str,
    game_id: &str,
    k: usize,
    shard_count: u32,
) -> Vec<(String, f64)> {
    let mut all_entries = Vec::new();

    for shard in 0..shard_count {
        let key = lb_key(tenant_id, game_id, shard);
        let entries = cache.zrevrange_withscores(&key, 0, k as isize - 1).await;
        all_entries.extend(entries);
    }

    // Sort by score descending
    all_entries.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));
    all_entries.truncate(k);
    all_entries
}

pub async fn get_approx_rank(
    cache: &Cache,
    tenant_id: &str,
    game_id: &str,
    player_id: &str,
    shard_count: u32,
) -> Option<usize> {
    let shard = shard_index(player_id, shard_count);
    let key = lb_key(tenant_id, game_id, shard);

    let score = cache.zscore(&key, player_id).await?;

    let mut higher_count: usize = 0;
    for s in 0..shard_count {
        let k = lb_key(tenant_id, game_id, s);
        let entries = cache.zrevrange_withscores(&k, 0, -1).await;
        for (_, entry_score) in &entries {
            if *entry_score > score {
                higher_count += 1;
            }
        }
    }

    Some(higher_count + 1)
}

pub async fn get_global_top_k(
    cache: &Cache,
    tenant_id: &str,
    k: usize,
    shard_count: u32,
) -> Vec<(String, f64)> {
    let mut all_entries = Vec::new();

    for shard in 0..shard_count {
        let key = global_key(tenant_id, shard);
        let entries = cache.zrevrange_withscores(&key, 0, k as isize - 1).await;
        all_entries.extend(entries);
    }

    all_entries.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));
    all_entries.truncate(k);
    all_entries
}

pub async fn rebuild_from_db(
    db: &sqlx::PgPool,
    cache: &Cache,
    tenant_id: &str,
    game_id: &str,
    shard_count: u32,
) -> AppResult<usize> {
    let rows: Vec<(String, i64)> = sqlx::query_as(
        "SELECT player_id::text, high_score FROM game_progress WHERE tenant_id = $1 AND game_id = $2 ORDER BY high_score DESC LIMIT 10000",
    )
    .bind(tenant_id)
    .bind(game_id)
    .fetch_all(db)
    .await?;

    let count = rows.len();
    for (pid, score) in &rows {
        update_score(cache, tenant_id, game_id, pid, *score as f64, shard_count).await;
    }
    Ok(count)
}
